Mapped several things across all the modelling classes and renderers

Sollace opened the pull request:
It's quite a lot, so I can't really cover every little thing I did in a description here, sorry!

Sollace committed:
Mapped CubeFace
field_3958 -> DIRECTION_LOOKUP
method_3163 -> getFace

Sollace committed:
WeightedBakedModel.ModelEntry -> WeightedBakedModel.Entry

Sollace committed:
Json ModelElement

asiekierka commented on mappings/net/minecraft/client/render/model/BakedModelManager.mapping:
```diff
@@ -1,11 +1,17 @@
+	FIELD field_5408 models Ljava/util/Map;
 	FIELD field_5409 spriteAtlas Lnet/minecraft/class_1059;
-	FIELD field_5410 blockStateMaps Lnet/minecraft/class_773;
+	FIELD field_5410 cache Lnet/minecraft/class_773;
```
I think that's insufficiently descriptive

Sollace replied:
Perhaps you're right. I double-checked and this field isn't a map as I'd remembered. It's a `BlockModels`.

The name comes from here:

```
    protected void apply(ModelLoader loader, ResourceManager manager, Profiler profiler) {
        profiler.startTick();
        profiler.push("upload");
        loader.upload(profiler);
        this.models = loader.getBakedModelMap();
        this.stateLookup = loader.getStateLookup();
        this.missingModel = this.models.get(ModelLoader.MISSING);
        profiler.swap("cache");
        this.cache.reload();
        profiler.pop();
        profiler.endTick();
    }
```

Though `cache` here could very well be referring to the method `reload()` as that proceeds to repopulate the internal `models` map, which could very well be named `cache` (as it is a cache of computed models).

Or the class itself could be named `BakedModelsCache` (as it only works with/contains `BakedModels` and serves as a buffer in front of the `BakedModelManager`/Factory).

It's less work just to name the field `blockModels` after the class, I suppose. :/

asiekierka commented on mappings/net/minecraft/client/render/model/BakedQuadFactory.mapping:
```diff
@@ -1,5 +1,79 @@
+	METHOD method_3465 getFactoryId (Lnet/minecraft/class_1086;Lnet/minecraft/class_2350;)I
+		ARG 0 rotation
+		ARG 1 direction
+	METHOD method_3466 registerUVMapping (Lnet/minecraft/class_1086;Lnet/minecraft/class_2350;Lnet/minecraft/class_796$class_797;)V
```
registerUvMapping

Sollace commented:
Rebased onto the current 1.14.4 branch to resolve some conflicts.

Runemoro commented on mappings/net/minecraft/client/render/model/BakedQuadFactory.mapping:
```diff
@@ -1,5 +1,79 @@
 CLASS net/minecraft/class_796 net/minecraft/client/render/model/BakedQuadFactory
+	CLASS class_797 UVMapper
```
`UvMapper`

Sollace replied:
:V

Runemoro commented on mappings/net/minecraft/client/render/model/BakedQuadFactory.mapping:
```diff
@@ -1,5 +1,79 @@
+		ARG 2 direction
+		ARG 3 corner
+		ARG 4 rotation
+	METHOD method_3456 getLightBrightness (Lnet/minecraft/class_2350;)F
```
Is `Light` necessary? What other kinds of brightnesses are there?

Sollace replied:
There is the base texture brightness and the directional ambient brightness that changes between different sides. This was specifically dealing with the latter.

liach commented on mappings/net/minecraft/client/render/model/MultipartUnbakedModel.mapping:
```diff
@@ -3,9 +3,9 @@ CLASS net/minecraft/class_816 net/minecraft/client/render/model/MultipartUnbaked
-		METHOD deserialize (Lcom/google/gson/JsonElement;Ljava/lang/reflect/Type;Lcom/google/gson/JsonDeserializationContext;)Ljava/lang/Object;
-			ARG 1 functionJson
-			ARG 2 unused
+		METHOD deserialize deserialize (Lcom/google/gson/JsonElement;Ljava/lang/reflect/Type;Lcom/google/gson/JsonDeserializationContext;)Ljava/lang/Object;
```
remove this mapping.
```suggestion
		METHOD deserialize (Lcom/google/gson/JsonElement;Ljava/lang/reflect/Type;Lcom/google/gson/JsonDeserializationContext;)Ljava/lang/Object;
```

liach reviewed (commented):
Fixes look nice. Dunno how portable they are to 1.15 given all these b3d changes.

Juuxel commented on mappings/net/minecraft/client/util/math/Vector4f.mapping:
```diff
@@ -17,3 +18,5 @@ CLASS net/minecraft/class_1162 net/minecraft/client/util/math/Vector4f
 	METHOD method_4957 getZ ()F
 	METHOD method_4958 getW ()F
+	METHOD method_4959 (Lnet/minecraft/class_1158;)V
+		ARG 1 quarternion
```
```suggestion
		ARG 1 quaternion
```

Sollace committed:
ModelLoader
ModelLoader.class_4455 -> ModelLoader.ModelDefinition (as hinted by Mojang)

Sollace committed:
BakedModel
deserialize -> deserialize (where's the synthetic?)

Sollace committed:
Vector4f

Sollace committed:
Resource reload listener

Sollace committed:
Mapped BakedQuadFactory

Sollace commented:
@modmuss50 @Juuxel @liach Update and rebased, and base branch changed to 19w44a!

This may need re-reviewing, as I saw there were quite a few differences in 1.15. (luckily, mostly deletions, but still)

liach commented on mappings/net/minecraft/client/render/model/BakedQuadFactory.mapping:
```diff
@@ -1,5 +1,59 @@
+		ARG 2 transformation
+	METHOD method_3456 getLightBrightness (Lnet/minecraft/class_2350;)F
+		ARG 1 direction
+	METHOD method_3457 getBrightnessInteger (Lnet/minecraft/class_2350;)I
```
Is the brightness integer the light map coordinate (block and sky light combination)?

Sollace replied:
No, I don't think it's the same as world light values. These are the 3D lighting effects applied to models, which vary per face.

`getLightBrightness` is the relative brightness for a given face by direction.
`getBrightnessInteger` is the result of `getLightBrightness` packed into an integer value.

```
    private int getBrightnessInteger(Direction direction) {
        int brightness = MathHelper.clamp((int)(this.getLightBrightness(direction) * 255), 0, 255);
        return 0xFF000000 | brightness << 16 | brightness << 8 | brightness;
    }
    
    private float getLightBrightness(Direction direction) {
        switch (direction) {
            case DOWN: return 0.5F;
            case UP: return 1F;
            case NORTH:
            case SOUTH: return 0.8F;
            case WEST:
            case EAST: return 0.6F;
            default: return 1F;
        }
    }
```

It's calculated and passed down as one of the components that are packed into the vertex buffer thing.

Here:
```
int brightness = shaded ? this.getBrightnessInteger(direction11) : -1;
```

And here:
```
        vertices[integer8] = Float.floatToRawIntBits(position.getX());
        vertices[integer8 + 1] = Float.floatToRawIntBits(position.getY());
        vertices[integer8 + 2] = Float.floatToRawIntBits(position.getZ());
        vertices[integer8 + 3] = brightness;
        vertices[integer8 + 4] = Float.floatToRawIntBits(sprite.getU(texture.getU(cornerIndex)));
        vertices[integer8 + 4 + 1] = Float.floatToRawIntBits(sprite.getV(texture.getV(cornerIndex)));
```

Runemoro replied:
This is called `light` everywhere else. It's the uv coordinates on the lightmap.

liach commented on mappings/net/minecraft/client/render/model/MultipartUnbakedModel.mapping:
```diff
@@ -3,9 +3,9 @@ CLASS net/minecraft/class_816 net/minecraft/client/render/model/MultipartUnbaked
-		METHOD deserialize (Lcom/google/gson/JsonElement;Ljava/lang/reflect/Type;Lcom/google/gson/JsonDeserializationContext;)Ljava/lang/Object;
-			ARG 1 functionJson
-			ARG 2 unused
+		METHOD deserialize deserialize (Lcom/google/gson/JsonElement;Ljava/lang/reflect/Type;Lcom/google/gson/JsonDeserializationContext;)Ljava/lang/Object;
```
Please include this suggested change.
```suggestion
		METHOD deserialize (Lcom/google/gson/JsonElement;Ljava/lang/reflect/Type;Lcom/google/gson/JsonDeserializationContext;)Ljava/lang/Object;
```

Sollace replied:
No. That displays as unmapped with a intermediate name in enigma.

liach replied:
That "displays" but shouldn't have a mapped entry here; it will not harm loading and may have trouble with matcher matching in the future.

Sollace replied:
But will the method actually be _named_ when using these mappings for mod development?

liach replied:
This breaks things. we had precedents.

liach commented on mappings/net/minecraft/client/render/model/json/ModelElement.mapping:
```diff
@@ -1,5 +1,9 @@
 CLASS net/minecraft/class_785 net/minecraft/client/render/model/json/ModelElement
 	CLASS class_786 Deserializer
+		METHOD deserialize deserialize (Lcom/google/gson/JsonElement;Ljava/lang/reflect/Type;Lcom/google/gson/JsonDeserializationContext;)Ljava/lang/Object;
```
Please include this suggested change.
```suggestion
		METHOD deserialize (Lcom/google/gson/JsonElement;Ljava/lang/reflect/Type;Lcom/google/gson/JsonDeserializationContext;)Ljava/lang/Object;
```

natanfudge replied:
why? lol

liach replied:
You are just not supposed to map this lol. In every other pr we don't include such duplicate names, especially for those screen class mappings

Runemoro replied:
How did this even get mapped? Enigma shouldn't generate names for these.

Sollace replied:
This mapping relates to the `deserialize` method on all the deserializers:

```java
        public ModelElement method_3406/*deserialize*/(JsonElement json, Type type, JsonDeserializationContext context) throws JsonParseException {
            JsonObject jsonObject5 = json.getAsJsonObject();
            Vector3f vector3f6 = this.deserializeFrom(jsonObject5);
            Vector3f vector3f7 = this.deserializeTo(jsonObject5);
            ModelRotation modelRotation8 = this.deserializeRotation(jsonObject5);
            Map<Direction, ModelElementFace> map9 = this.deserializeFacesValidating(context, jsonObject5);
            if (jsonObject5.has("shade") && !JsonHelper.hasBoolean(jsonObject5, "shade")) {
                throw new JsonParseException("Expected shade to be a Boolean");
            }
            boolean boolean10 = JsonHelper.getBoolean(jsonObject5, "shade", true);
            return new ModelElement(vector3f6, vector3f7, map9, modelRotation8, boolean10);
        }
```

I think there's some bug in Enigma, though, because based on this signature this line in the mappings should look like this:

```
METHOD method_3406 deserialize (Lcom/google/gson/JsonElement;Ljava/lang/reflect/Type;Lcom/google/gson/JsonDeserializationContext;)Lnet/minecraft/class_785;

liach commented on mappings/net/minecraft/client/render/model/json/ModelElementTexture.mapping:
```diff
@@ -1,6 +1,6 @@
 CLASS net/minecraft/class_787 net/minecraft/client/render/model/json/ModelElementTexture
 	CLASS class_788 Deserializer
-		METHOD deserialize (Lcom/google/gson/JsonElement;Ljava/lang/reflect/Type;Lcom/google/gson/JsonDeserializationContext;)Ljava/lang/Object;
+		METHOD deserialize deserialize (Lcom/google/gson/JsonElement;Ljava/lang/reflect/Type;Lcom/google/gson/JsonDeserializationContext;)Ljava/lang/Object;
```
Please include this suggested change.
```suggestion
		METHOD deserialize (Lcom/google/gson/JsonElement;Ljava/lang/reflect/Type;Lcom/google/gson/JsonDeserializationContext;)Ljava/lang/Object;
```

liach commented on mappings/net/minecraft/client/util/math/Vector4f.mapping:
```diff
@@ -24,7 +26,9 @@ CLASS net/minecraft/class_1162 net/minecraft/client/util/math/Vector4f
-	METHOD method_4954 multiplyXyz (Lnet/minecraft/class_1160;)V
-		ARG 1 other
+	METHOD method_4954 multiply (Lnet/minecraft/class_1160;)V
+		ARG 1 vector
```
Is multiplying x by x1, y by y1, and z by z1 of the vector a multiplication?

Sollace replied:
... yes. What would make that _not_ a multiplication?

liach replied:
A vector multiplication means a dot or cross product. This is neither.

On a side note, this 4-d vector serves like a https://en.wikipedia.org/wiki/Homogeneous_coordinates for a 3d space.

Sollace replied:
Alrighty... Is there a term relating to multiplying vectors by component then?

Runemoro replied:
Yes, "inner product".

Runemoro replied:
https://en.m.wikipedia.org/wiki/Hadamard_product_(matrices), but I would prefer `multiplyComponentwise`.

Sollace replied:
Hm. I was hoping there was a name I was just not remembering (like dotProduct, scalarProduct?). `multiplyComponentWise` looks like it will work.

Runemoro replied:
> multiplyComponentWise 

Lowercase `w`. "[componentwise](https://en.wiktionary.org/wiki/componentwise)" is an adjective. "component wise" means a component that's somehow wise, and in the wrong word order.

> was hoping there was a name

There are three. Read the first sentence (or even the title in the link itself!) of the Wikipedia article I linked above. "Hadamard product", "Schur product", and "entrywise product" (though for vectors, we refer to their entries as components).

liach commented:
@modmuss50 Can you dismiss the review from asiekierka?

Sollace commented:
Side note:

This is 1.14, so I don't know if it applies to 1.15, but as I've been working on modelling classes I've notices this in `Box`:

```
   public void render(BufferBuilder bufferBuilder_1, float float_1) {
      Quad[] var3 = this.polygons;
      int var4 = var3.length;

      for(int var5 = 0; var5 < var4; ++var5) {
         Quad quad_1 = var3[var5];
         quad_1.render(bufferBuilder_1, float_1);
      }
   }
```
The name `render` is a bit misleading, as they're only called once by the Cuboid to compile its call list.

Any votes to change it to:
`Box.render` -> `Box.compile`
`Quad.render` -> `Quad.compile`

Runemoro commented:
@Sollace None of those exist in 1.15.

Sollace commented:
> 
> 
> @Sollace None of those exist in 1.15. Also, if you didn't already, could you open your PR in Enigma and do "File > Drop invalid mappings"? I think that many of the things you mapped in your PR are gone in 1.15.

Sure. I thought I had already, but maybe that was only the other PR.

liach commented:
@Sollace please update this pr as well.

Sollace commented:
> 
> 
> @Sollace None of those exist in 1.15. Also, if you didn't already, could you open your PR in Enigma and do "File > Drop invalid mappings"? I think that many of the things you mapped in your PR are gone in 1.15.

I did a check for invalid mappings and enigma didn't find any.

@liach Rebased and updates.

liach commented:
Mind fixing the merge conflicts?

Sollace commented:
> 
> 
> Mind fixing the merge conflicts?

Yeah, rebasing again... -_-

Did someone merge something?

liach commented:
Turns out I accidentally mapped the state model group look up maps

liach commented on mappings/net/minecraft/client/render/model/BakedModelManager.mapping:
```diff
@@ -1,17 +1,17 @@
-	METHOD method_21611 shouldRerender (Lnet/minecraft/class_2680;Lnet/minecraft/class_2680;)Z
-		ARG 1 old
-		ARG 2 updated
+	METHOD method_21611 statesChanged (Lnet/minecraft/class_2680;Lnet/minecraft/class_2680;)Z
```
This returns if the two states have different models rather than if the states "have changed" (too ambiguous imo)

Sollace replied:
Alright. I don't think `shouldRerender` is accurate either. I'll try for something better once I get Enigma booted up.

Sollace committed:
Back to this I guess.

Sollace committed:
These mappings were lost in the rebase. :/

Sollace commented:
It seems like Vector4f got swapped with some other class during the update/rebase.

I remember the methods dealing with the W parameter being missing and having to remove them during the first rebase. Somehow they're back now, so I've had to just redo those mappings.

Also applied the suggestion from @Runemoro about `multiplyComponentwise`.

Sollace committed:
getLightBrightness -> getRelativeDirectionalBrightness
getBrightnessInteger -> getLightmapCoordinate

Sollace committed:
Mapped some incidental names relating to Sprite

Sollace commented:
Can someone check @asiekierka 's request? It's been over 3 months.

liach commented:
@Sollace Please update the base branch to 19w46b.

Sollace committed:
Vector4f.method_4929 -> Vector4f.viewboxMatrix

Sollace committed:
field_20848 -> BLOCK_DESTRUCTION_STAGES
BLOCK_BREAKING_STAGES -> BLOCK_DESTRUCTION_STAGE_TEXTURES

Sollace commented:
Rebased and updated!

Notable changes:
Matrix4f.method_4929 -> viewboxMatrix 
ModelLoader.field_20848 -> BLOCK_DESTRUCTION_STAGES
ModelLoader.BLOCK_BREAKING_STAGES ->BLOCK_DESTRUCTION_STAGE_TEXTURES
